python/cufinufft/tests/test_simple.py

```diff
@@ -11,77 +11,70 @@
 
 DTYPES = [np.float32, np.float64]
 SHAPES = [(64,), (64, 64), (64, 64, 64)]
+N_TRANS = [(), (1,), (2,)]
 MS = [256, 1024, 4096]
 TOLS = [1e-2, 1e-3]
 OUTPUT_ARGS = [False, True]
 
 @pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize("n_trans", N_TRANS)
+@pytest.mark.parametrize("shape", SHAPES)
 @pytest.mark.parametrize("M", MS)
 @pytest.mark.parametrize("tol", TOLS)
 @pytest.mark.parametrize("output_arg", OUTPUT_ARGS)
-def test_nufft3d1(dtype, M, tol, output_arg):
-    shape = (64, 64, 64)
-
+def test_simple_type1(dtype, shape, n_trans, M, tol, output_arg):
     real_dtype = dtype
     complex_dtype = utils._complex_dtype(dtype)
 
     dim = len(shape)
 
-    k = utils.gen_nu_pts(M, dim=dim).astype(real_dtype)
-    c = utils.gen_nonuniform_data(M).astype(complex_dtype)
+    fun = {1: cufinufft.nufft1d1,
+           2: cufinufft.nufft2d1,
+           3: cufinufft.nufft3d1}[dim]
+
+    k, c = utils.type1_problem(dtype, shape, M, n_trans=n_trans)
 
     k_gpu = gpuarray.to_gpu(k)
     c_gpu = gpuarray.to_gpu(c)
 
     if output_arg:
-        fk_gpu = gpuarray.GPUArray(shape, dtype=complex_dtype)
-        cufinufft.nufft3d1(*k_gpu, c_gpu, out=fk_gpu, eps=tol)
+        fk_gpu = gpuarray.GPUArray(n_trans + shape, dtype=complex_dtype)
+        fun(*k_gpu, c_gpu, out=fk_gpu, eps=tol)
     else:
-        fk_gpu = cufinufft.nufft3d1(*k_gpu, c_gpu, shape, eps=tol)
+        fk_gpu = fun(*k_gpu, c_gpu, shape, eps=tol)
 
     fk = fk_gpu.get()
 
-    ind = int(0.1789 * np.prod(shape))
-
-    fk_est = fk.ravel()[ind]
-    fk_target = utils.direct_type1(c, k, shape, ind)
-
-    type1_rel_err = np.abs(fk_target - fk_est) / np.abs(fk_target)
-
-    assert type1_rel_err < 10 * tol
+    utils.verify_type1(k, c, fk, tol)
 
 
 @pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize("shape", SHAPES)
+@pytest.mark.parametrize("n_trans", N_TRANS)
 @pytest.mark.parametrize("M", MS)
 @pytest.mark.parametrize("tol", TOLS)
 @pytest.mark.parametrize("output_arg", OUTPUT_ARGS)
-def test_nufft3d2(dtype, M, tol, output_arg):
-    shape = (64, 64, 64)
-
+def test_simple_type2(dtype, shape, n_trans, M, tol, output_arg):
     real_dtype = dtype
     complex_dtype = utils._complex_dtype(dtype)
 
     dim = len(shape)
 
-    k = utils.gen_nu_pts(M, dim=dim).astype(real_dtype)
-    fk = utils.gen_uniform_data(shape).astype(complex_dtype)
+    fun = {1: cufinufft.nufft1d2,
+           2: cufinufft.nufft2d2,
+           3: cufinufft.nufft3d2}[dim]
+
+    k, fk = utils.type2_problem(dtype, shape, M, n_trans=n_trans)
 
     k_gpu = gpuarray.to_gpu(k)
     fk_gpu = gpuarray.to_gpu(fk)
 
     if output_arg:
-        c_gpu = gpuarray.GPUArray((M,), dtype=complex_dtype)
-        cufinufft.nufft3d2(*k_gpu, fk_gpu, out=c_gpu)
+        c_gpu = gpuarray.GPUArray(n_trans + (M,), dtype=complex_dtype)
```

Comment:
likewise as a human I can't follow this ntrans + (M,)   why?    fun(*k_gpu, ... ,...) why?
Maybe just bit more comments for non-pythonistas.

Reply:
So `n_trans` represents the batch shape, which can be `()` (a single transform), `(1,)` (a single transform, but given as an array with first dimension equal to one), or `(n,)` (`n` different transforms).

What this allows us to do is to generate the output array of the correct size by adding the batch shape tuple `n_trans` together with the number of non-uniform points `M`. So in the three above cases we'd get array shapes of `() + (M,) = (M,)`, `(1,) + (M,) = (1, M)`, and `(n,) + (M,) = (n, M),` respectively.

Will add some comments here and in other spots.

Reply:
Done.